sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/planning/QueryPlanner.scala

```diff
@@ -47,17 +55,27 @@ abstract class QueryPlanner[PhysicalPlan <: TreeNode[PhysicalPlan]] {
   /** A list of execution strategies that can be used by the planner */
   def strategies: Seq[GenericStrategy[PhysicalPlan]]
 
-  /**
-   * Returns a placeholder for a physical plan that executes `plan`. This placeholder will be
-   * filled in automatically by the QueryPlanner using the other execution strategies that are
-   * available.
-   */
-  protected def planLater(plan: LogicalPlan): PhysicalPlan = this.plan(plan).next()
-
   def plan(plan: LogicalPlan): Iterator[PhysicalPlan] = {
     // Obviously a lot to do here still...
-    val iter = strategies.view.flatMap(_(plan)).toIterator
+    val iter = strategies.view.flatMap(_(plan)).toIterator.flatMap { physicalPlan =>
+      val placeholders = collectPlaceholders(physicalPlan)
+
+      (Iterator(physicalPlan) /: placeholders.toIterator) {
+        case (physicalPlans, (placeholder, logicalPlan)) =>
+          val children = this.plan(logicalPlan)
+          physicalPlans.flatMap { physicalPlan =>
+            children.map { child =>
+              physicalPlan.transformUp {
+                case `placeholder` => child
+              }
+            }
+          }
+      }
+    }
+    // TODO: We will need to prune bad plans to prevent from combinatorial explosion.
```

Comment:
We are safe for now though as long as we only call `next()` once right?  Is there some way to test that its doing the right thing?